File: src/project_images.py
"""Bounded local image preparation. Pillow is loaded only for image operations."""
import base64
import io
import warnings
from pathlib import Path
from engine import fail, digest


def prepare_images(engine):
    if engine.project.get('target') != 'local':
        fail('LOCAL_ONLY', 'This version supports local projects only.')
    paths = engine.args.get('paths', [])
    if not isinstance(paths, list) or not 1 <= len(paths) <= 4 or len(set(paths)) != len(paths):
        fail('INVALID_INPUT', 'Supply 1 to 4 distinct image paths.')
    result = []
    for index, rel in enumerate(paths):
        supplied = Path(rel)
        if supplied.is_absolute():
            from local_images import local_path
            p = local_path(rel)
            # Keep historical relative labels for images within the selected project.
            for root in (Path(engine.project['root']), engine.root):
                try:
                    rel = supplied.relative_to(root).as_posix()
                    break
                except ValueError:
                    continue
        else:
            if not engine.args.get('project_id'):
                fail('LOCAL_PATH_REQUIRED', 'Without project_id, supply absolute local image paths.')
            p = engine.path(rel)
        if not p.is_file():
            fail('NOT_FOUND', 'Image not found: ' + rel)
        if p.suffix.lower() not in {'.png', '.jpg', '.jpeg'}:
            fail('UNSUPPORTED_IMAGE', 'Only PNG and JPEG are supported: ' + rel)
        with p.open('rb') as source:
            raw = source.read(32 * 1024 * 1024 + 1)
        if len(raw) > 32 * 1024 * 1024:
            fail('IMAGE_TOO_LARGE', 'Image exceeds 32 MiB: ' + rel)
        result.append(prepare_image(raw, p.suffix, rel, index + 1))
    return {'images': result}


def prepare_image(raw, suffix, rel, index):
    try:
        from PIL import Image, ImageOps
    except ImportError:
        fail('IMAGE_DEPENDENCY_MISSING', 'Install Pillow in the configured Bridge Python environment.')
    if len(raw) > 32 * 1024 * 1024:
        fail('IMAGE_TOO_LARGE', 'Image exceeds 32 MiB: ' + rel)
    if suffix.lower() not in {'.png', '.jpg', '.jpeg'}:
        fail('UNSUPPORTED_IMAGE', 'Only PNG and JPEG are supported: ' + rel)
    try:
        with warnings.catch_warnings():
            warnings.simplefilter('error', Image.DecompressionBombWarning)
            with Image.open(io.BytesIO(raw)) as source:
                expected = 'PNG' if suffix.lower() == '.png' else 'JPEG'
                if source.format != expected or getattr(source, 'n_frames', 1) != 1:
                    fail('UNSUPPORTED_IMAGE', 'File must be a single PNG or JPEG matching its extension: ' + rel)
                original_size = source.size
                if source.width * source.height > 40_000_000:
                    fail('IMAGE_TOO_LARGE', 'Image exceeds 40 million pixels: ' + rel)
                source.load()
                rotated = source.getexif().get(274, 1) != 1
                image = ImageOps.exif_transpose(source)
                image.thumbnail((2560, 2560), Image.Resampling.LANCZOS)
                resized = image.size != original_size and not (rotated and image.size == original_size[::-1])
                output = raw
                recompressed = False
                if resized or rotated or len(raw) > 768 * 1024:
                    for attempt in range(10):
                        stream = io.BytesIO()
                        if expected == 'JPEG':
                            image.convert('RGB').save(stream, format='JPEG', quality=85, optimize=True)
                        else:
                            image.save(stream, format='PNG', optimize=True)
                        output = stream.getvalue()
                        if len(output) <= 768 * 1024:
                            break
                        image = image.resize((max(1, int(image.width * .75)), max(1, int(image.height * .75))), Image.Resampling.LANCZOS)
                        resized = True
                    recompressed = True
                if len(output) > 768 * 1024:
                    fail('IMAGE_TOO_LARGE', 'Prepared image exceeds transport limit: ' + rel)
                return {'index': index, 'path': rel, 'mime_type': 'image/png' if expected == 'PNG' else 'image/jpeg',
                    'source_bytes': len(raw), 'bytes': len(output), 'source_sha256': digest(raw), 'sha256': digest(output),
                    'original_width': original_size[0], 'original_height': original_size[1], 'width': image.width, 'height': image.height,
                    'resized': resized, 'orientation_corrected': rotated, 'recompressed': recompressed,
                    'base64': base64.b64encode(output).decode('ascii')}
    except Exception as error:
        if hasattr(error, 'code'):
            raise
        fail('INVALID_IMAGE', 'Cannot decode image safely: ' + rel)

File: src/project-images.mjs
import { readFile, writeFile, mkdir, open } from 'node:fs/promises';
import { randomUUID } from 'node:crypto';
import path from 'node:path';
import { z } from 'zod';
import { config, invoke } from './bridge.mjs';

export const imageViewerUri = 'ui://x14/project-images-v4.html';
const identity = z.string().uuid();
export function registerProjectImages(server, suppliedConfig) {
  const getConfig = async () => suppliedConfig || await config();
  server.registerResource('project-images', imageViewerUri, { mimeType: 'text/html;profile=mcp-app' }, async () => {
    const [html, js] = await Promise.all([readFile(new URL('./ui/project-images.html', import.meta.url), 'utf8'), readFile(new URL('./ui/project-images.js', import.meta.url), 'utf8')]);
    return { contents: [{ uri: imageViewerUri, mimeType: 'text/html;profile=mcp-app', text: html.replace('/* INLINE_SCRIPT */', js), _meta: {
      ui: { prefersBorder: true, csp: { connectDomains: [], resourceDomains: [] } },
      'openai/widgetCSP': { connect_domains: [], resource_domains: [] },
      'openai/widgetDescription': 'Automatically prepares local project images for the next model turn. Initially say only that images are being prepared; inspect images and continue the original authorized task after the automatic follow-up. Do not report a vision failure while preparation is pending.'
    } }] };
  });
  server.registerTool('view_project_images', {
    title: '查看本地图片',
    description: 'View 1–4 local PNG/JPEG images anywhere on this computer using ChatGPT vision. Absolute local paths need no project_id; relative paths require a local project_id. If location is unknown use discover_local_images. Source limits: 32 MiB and 40 million pixels per image; delivery optimized to 2560px / 768 KiB. Pass original task context and remaining authorized work. Initially say preparing; after automatic follow-up inspect images and resume the task. Do not repeat pending requests. Network drives and SSH are not supported.',
    inputSchema: { project_id: z.string().max(100).optional(), paths: z.array(z.string().min(1).max(1000)).min(1).max(4), question: z.string().min(1).max(6000) },
    outputSchema: { run_id: identity.optional(), status: z.enum(['waiting_for_widget', 'error']), error: z.string().optional(), project_id: z.string().optional(), question: z.string().optional(), images: z.array(z.object({ index: z.number(), path: z.string(), mime_type: z.string(), source_bytes: z.number(), bytes: z.number(), source_sha256: z.string(), sha256: z.string(), original_width: z.number(), original_height: z.number(), width: z.number(), height: z.number(), resized: z.boolean(), orientation_corrected: z.boolean(), recompressed: z.boolean() })).optional() },
    annotations: { readOnlyHint: true, destructiveHint: false, idempotentHint: false, openWorldHint: false },
    _meta: { ui: { resourceUri: imageViewerUri }, 'openai/outputTemplate': imageViewerUri, 'openai/toolInvocation/invoking': '正在读取项目图片', 'openai/toolInvocation/invoked': '正在准备图片，稍后自动分析' }
  }, async args => {
    try {
      const c = await getConfig();
      const project = c.projects.find(p => p.id === args.project_id);
      if (args.project_id && !project) throw new Error('UNAUTHORIZED: unknown project_id');
      if (project && project.target !== 'local') throw new Error('LOCAL_ONLY: 当前仅支持本地项目图片');
      const result = await invoke('prepare_images', args, c);
      if (!result.ok) throw new Error(result.error.code + ': ' + result.error.message);
      const run_id = randomUUID(), token = randomUUID();
      const images = result.result.images.map(({ base64, ...info }) => info);
      const folder = path.join(c.state, 'image-deliveries');
      await mkdir(folder, { recursive: true });
      await writeFile(path.join(folder, run_id + '.json'), JSON.stringify({ token, created_at: Date.now() }), { flag: 'wx' });
      const data = { run_id, status: 'waiting_for_widget', project_id: args.project_id, question: args.question, images };
      return { content: [{ type: 'text', text: '正在准备图片，稍后自动分析。请等待卡片自动发起后续回答，不要重复调用。' }], structuredContent: data,
        _meta: { 'x14/images': { token, base64: result.result.images.map(i => i.base64) } }, isError: false };
    } catch (error) { return { content: [{ type: 'text', text: String(error.message) }], structuredContent: { status: 'error', error: String(error.message) }, isError: true }; }
  });
  // Only the UI receives the per-run token. Exclusive creation also works across processes.
  server.registerTool('claim_image_delivery', {
    title: '图片提交去重（内部）', description: 'Widget-only one-time claim before uploading a prepared image batch. Not a user-facing image reader.',
    inputSchema: { run_id: identity, token: identity }, outputSchema: { granted: z.boolean(), reason: z.string() },
    annotations: { readOnlyHint: false, destructiveHint: false, idempotentHint: true, openWorldHint: false },
    _meta: { ui: { visibility: ['app'] }, 'openai/visibility': 'private', 'openai/widgetAccessible': true }
  }, async ({ run_id, token }) => {
    const c = await getConfig(), folder = path.join(c.state, 'image-deliveries');
    let data;
    try {
      const record = JSON.parse(await readFile(path.join(folder, run_id + '.json'), 'utf8'));
      if (record.token !== token || Date.now() - record.created_at > 86400000) throw new Error('invalid');
      try { const handle = await open(path.join(folder, run_id + '.claimed'), 'wx'); await handle.close(); data = { granted: true, reason: 'claimed' }; }
      catch (error) { if (error.code !== 'EEXIST') throw error; data = { granted: false, reason: 'already_claimed; do not upload or send another follow-up' }; }
    } catch { data = { granted: false, reason: 'invalid_or_expired_delivery' }; }
    return { content: [{ type: 'text', text: JSON.stringify(data) }], structuredContent: data };
  });
}
